### ISSUE TITLE
false positive of `let_unit_value` on async fn arguments

### ISSUE THREAD

**Issue by udoprog** [rust-lang/rust-clippy#11502]
### Summary

False positive of `let_unit_value` on async fn arguments.

### Lint Name

let_unit_value

### Reproducer

I tried this code:

```rust
pub async fn hello(_: ()) {
}
```

I saw this happen:

```
warning: this let-binding has unit value
 --> src/lib.rs:1:20
  |
1 | pub async fn hello(_: ()) {
  |                    ^ help: omit the `let` binding: `_;`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_unit_value
  = note: `#[warn(clippy::let_unit_value)]` on by default
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021)

### Version

```text
rustc 1.74.0-nightly (8142a319e 2023-09-13)
binary: rustc
commit-hash: 8142a319ed5c1d1f96e5a1881a6546e463b77c8f
commit-date: 2023-09-13
host: x86_64-pc-windows-msvc
release: 1.74.0-nightly
LLVM version: 17.0.0
```

**Comment by Centri3** [comment]
This is desugared to something like:
```rust
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;

async fn hello(a: ()) ->  ... {
    let a = a;
    {
        let _t = { };
        _t
    }
}

```
So basically, the `unit_types` lint group just needs a `from_expansion` check. We actually already have this for the local's pattern, but that doesn't seem to be enough (we should check the local's span instead; maybe it copies the tokens of the pattern for that argument, or smth)
